web3/utils/contracts.py

```diff
@@ -70,36 +77,54 @@ def find_matching_event_abi(abi, event_name=None, argument_names=None):
 
 
 def find_matching_fn_abi(abi, fn_identifier=None, args=None, kwargs=None):
+    args = args or tuple()
+    kwargs = kwargs or dict()
     filters = []
+    num_arguments = len(args) + len(kwargs)
 
-    if fn_identifier:
-        if fn_identifier is FallbackFn:
-            return get_fallback_func_abi(abi)
-        elif is_text(fn_identifier):
-            filters.append(functools.partial(filter_by_name, fn_identifier))
-        else:
-            raise TypeError("Unsupported function identifier")
-
-    if args is not None or kwargs is not None:
-        if args is None:
-            args = tuple()
-        if kwargs is None:
-            kwargs = {}
-
-        num_arguments = len(args) + len(kwargs)
-        filters.extend([
-            functools.partial(filter_by_argument_count, num_arguments),
-            functools.partial(filter_by_encodability, args, kwargs),
-        ])
+    if fn_identifier is FallbackFn:
+        return get_fallback_func_abi(abi)
 
-    function_candidates = pipe(abi, *filters)
+    if not is_text(fn_identifier):
+        raise TypeError("Unsupported function identifier")
 
+    # Exact match
+    name_filter = functools.partial(filter_by_name, fn_identifier)
+    arg_count_filter = functools.partial(filter_by_argument_count, num_arguments)
+    encoding_filter = functools.partial(filter_by_encodability, args, kwargs)
+    filters.extend([
+        name_filter,
+        arg_count_filter,
+        encoding_filter,
+    ])
+    function_candidates = pipe(abi, *filters)
     if len(function_candidates) == 1:
         return function_candidates[0]
-    if not function_candidates:
-        raise ValueError("No matching functions found")
+    # No exact match
     else:
-        raise ValueError("Multiple functions found")
+        matching_identifiers = name_filter(abi)
+        matching_function_signatures = [abi_to_signature(func) for func in matching_identifiers]
+        arg_count_matches = len(arg_count_filter(matching_identifiers))
+        encoding_matches = len(encoding_filter(matching_identifiers))
+        if arg_count_matches == 0:
+            diagnosis = "\nFunction invocation failed due to improper number of arguments."
+        elif encoding_matches == 0 or encoding_matches > 1:
+            diagnosis = "\nFunction invocation failed due to improper argument encoding."
```

Comment:
So lets say there are two functions: `func1(string)` and `func1(bytes)`. All calls to the function with a string will fail with the message "Function invocation failed due to improper argument encoding", right? We probably want different messages for `encoding_matches == 0` and `encoding_matches > 1`.